### ISSUE TITLE
Can't send an error code or message when rejecting a connection due to auth failure

### ISSUE THREAD

**Issue by pierznj** [websockets/ws#1745]
- ☑️  I've searched for any related issues and avoided creating a duplicate
      issue.

#### Description

Your example of client authentication here: https://github.com/websockets/ws#client-authentication sends some text via the Node socket (socket.write('HTTP/1.1 401 Unauthorized\r\n\r\n');) and then calls socket.destroy(), but the text sent using socket.write is never seen by the client AFAICT. Thus the client never knows the reason for the socket connection rejection. 

#### Reproducible in:

- version: 7.2.3
- Node.js version(s): 10
- OS version(s): Linux

#### Steps to reproduce:

1. Implement an auth mechanism per the example

2. Send a socket connection request with bad credentials

#### Expected result:

Socket receives a message with 401 Unauthorized

#### Actual result:

Socket is closed without an informative error code or message

**Comment by lpinca** [comment]
What you mean with error code or message? The connection is still not upgraded, what is written to the socket is an HTTP response. If you want to add headers or a body with JSON response you can do that.

**Comment by pierznj** [comment]
Thanks - it makes sense that the response is HTTP, and I looked for evidence of a successful 401  response in the Chrome network tab response headers for that socket request, but there is no sign of it. The response headers look like:

HTTP/1.1 101 Switching Protocols
Upgrade: websocket
Connection: Upgrade
Sec-WebSocket-Accept: <some id>

**Comment by lpinca** [comment]
Yes, that is to complete the handshake. You don't want to complete it if the authentication fails and that's why the example use a minimal 401 response.

The only gotcha is that you have a raw socket.

**Comment by pierznj** [comment]
Hmm, so I'm still left unclear as to how to detect the auth failure on the client. I have no idea where that 401 header is going or how to catch it in the browser.

**Comment by lpinca** [comment]
Unfortunately the WebSocket interface in the browser does not allow you to read the HTTP response of the opening handshake so there is no proper way. It will probably call `websocet.onerror` and certainly `websocet.onclose` with 1006.

**Comment by lpinca** [comment]
Closing as answered. Discussion can continue if needed.

**Comment by pierznj** [comment]
It does not call onerror, which leaves me with handling the onclose with code 1006 - not exactly a satisfactory situation since 1006 could occur for other reasons than an auth failure. I know verifyClient is discouraged, but would using it give me better options here? Or does that also abandon the handshake if verification fails?

**Comment by lpinca** [comment]
`verifyClient()` works exactly is in the same way. I tried it on latest chrome and it calls `ws.onerror`. Another way would be to not abort the handshake, wait for the connection to be established and then close the websocket on the server after sending a custom made message or closing with a custom close code but I wouldn't do it.

**Comment by ngreenwood6** [comment]
@pierznj I have found a solution to the problem. While I also believe the library should support this as well, I believe this workaround should work for you. Just for reference, the reason this needs to be added to the library is because of re-connects. For example, I have my client trying to reconnect with a back off (starts at 1 second and multiplies by 1.5x until it reaches a total of 30 seconds). Now when I receive a 401 obviously it's not trying to re-connect because of a network issue and rather an authentication issue. In this case, I want it to stop trying to re-connect because the user obviously has to re-authenticate. Solution is here:

```
socket.on('unexpected-response', (req, res) => {
      this.socket.emit('close', res.statusCode);
})

.socket.on('close', (code) => {
      // try to reconnect
      if (code !== 401) this.reconnect();
})
```

So basically what I am doing there is catching the `unexpected-response` event from the library (go to the source websocket.js file and search for it). That event actually sends the response from the server, which you can then get the status code from. In my case, the close event was not getting triggered, so I trigger it manually sending it the status code. Now in the `close` event, you can listen for your status code and detect if it is a 401.

Hoping this helps anyone in the future.

**Comment by snakepoongmail** [comment]
> `verifyClient()` works exactly is in the same way. I tried it on latest Chrome and it calls `ws.onerror`.
> 
> Another way would be to not abort the handshake, wait for the connection to be established and then close the websocket on the server after sending a custom made message or closing with a custom close code but I wouldn't do it.

Hello @lpinca could you explain more the reason why this is not a good option? Is it security concern base or other reason? Thx.
